### PR TITLE
Refactor heartbeat

### PR DESCRIPTION
**Pull request by ganmacs**
**Which issue(s) this PR fixes**: 
no

**What this PR does / why we need it**: 

Refactor heartbeat code in out_forward.
Merged duplicated code and use more specify method name.

**Docs Changes**:

no needed

**Release Note**: 

no needed

### CODE REVIEW

File: lib/fluent/plugin/out_forward.rb [hunk 1]
```diff
@@ -177,18 +177,14 @@ def configure(conf)
...
+      if @dns_round_robin && @heartbeat_type == :udp
+        raise Fluent::ConfigError, "forward output heartbeat type must be 'transport' or 'none' to use dns_round_robin option"
       end
 
       if @transport == :tls
         # socket helper adds CA cert or signed certificate to same cert store internally so unify it in this place.
         if @tls_cert_path && !@tls_cert_path.empty?
           @tls_ca_cert_path = @tls_cert_path
-        end
-        if @tls_ca_cert_path && !@tls_ca_cert_path.empty?
```
Comment (repeatedly) [inline lib/fluent/plugin/out_forward.rb]: Removing this `if` is safe?

File: lib/fluent/plugin/out_forward.rb [hunk 2]
```diff
@@ -401,12 +394,15 @@ def on_timer
...
-      if node = @nodes.find {|n| n.sockaddr == sockaddr }
+    def on_udp_heatbeat_response_recv(data, sock)
+      sockaddr = Socket.pack_sockaddr_in(sock.remote_port, sock.remote_host)
+      if node = @nodes.find { |n| n.sockaddr == sockaddr }
         # log.trace "heartbeat arrived", name: node.name, host: node.host, port: node.port
         if node.heartbeat
           @load_balancer.rebuild_weight_array(@nodes)
         end
+      else
+        log.warn("Unknown heartbeat response received host: #{soc.remote_host}, port: #{sock.remote_port}")
```
Comment (repeatedly) [inline lib/fluent/plugin/out_forward.rb]: `received from #{sock.remote_host}:#{sock.remote_port}` is more readable for me.

File: lib/fluent/plugin/out_forward.rb (hunk 1 above)
Comment (ganmacs) [reply lib/fluent/plugin/out_forward.rb]: I found that `@tls_ca_cert_path` is defined in cofig_params. this change is wrong. reverted it.
https://github.com/fluent/fluentd/pull/2503/commits/4426b028616a4d1bc4c6245d63c003aa94117dde

File: lib/fluent/plugin/out_forward.rb (hunk 2 above)
Comment (ganmacs) [reply lib/fluent/plugin/out_forward.rb]: https://github.com/fluent/fluentd/pull/2503/commits/efa756543e4e982145b4c17bcdaec007bf2520ea

File: lib/fluent/plugin/out_forward.rb [hunk 3]
```diff
@@ -401,12 +396,15 @@ def on_timer
...
-      if node = @nodes.find {|n| n.sockaddr == sockaddr }
+    def on_udp_heatbeat_response_recv(data, sock)
+      sockaddr = Socket.pack_sockaddr_in(sock.remote_port, sock.remote_host)
+      if node = @nodes.find { |n| n.sockaddr == sockaddr }
         # log.trace "heartbeat arrived", name: node.name, host: node.host, port: node.port
         if node.heartbeat
           @load_balancer.rebuild_weight_array(@nodes)
         end
+      else
+        log.warn("Unknown heartbeat response received from #{soc.remote_host}:#{sock.remote_port}")
```
Comment (repeatedly) [inline lib/fluent/plugin/out_forward.rb]: Is `soc` typo?

Comment (ganmacs) [reply lib/fluent/plugin/out_forward.rb]: my bad… fixed https://github.com/fluent/fluentd/pull/2503/commits/98417e64f7d8c350c9acadbc775c0d6786122e67

**Pull request merged**